### PR TITLE
New-AzureRmVm SimpleParameterSet, Strategies

### PR DESCRIPTION
**Pull request by sergey-shandar**
New-AzureRmVm SimpleParameterSet, Strategies

## Description
- New-AzureRmVm SimpleParameterSet
- Strategies library

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/Common/Commands.Common/AzureLongRunningJob.cs [hunk 1]
```diff
@@ -191,11 +191,13 @@ public override string StatusMessage
                 returnValue.MyInvocation.BoundParameters.Add(parameter.Key, parameter.Value);
             }
 
-            foreach (var field in returnType.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly))
+
+            foreach (var field in returnType.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
```
Comment (markcowl) [inline src/Common/Commands.Common/AzureLongRunningJob.cs]: I mistakenly changed this line in my PR - if you pull from my PR into release-5.1.0, this line is unchanged

File: src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Azure.Commands.Common.Strategies.UnitTest
+{
+    [TestClass]
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs]: These should be xunit tests, and we will also need entries for them in the testmap

File: src/ResourceManager/Common/Commands.Common.Strategies/Commands.Common.Strategies.csproj [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>Microsoft.Azure.Commands.Common.Strategies</RootNamespace>
+    <AssemblyName>Microsoft.Azure.Commands.Common.Strategies</AssemblyName>
+    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>full</DebugType>
+    <Optimize>false</Optimize>
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies/Commands.Common.Strategies.csproj]: Generally, for common code, the output path should be ```bin/<Config>``` and the dll is copied over by project references.

Also, you need a release target that does signing and a singing key (copy from any of the common assemblies)

File: tools/UpdateModules.ps1 [hunk 4]
```diff
@@ -50,7 +50,7 @@ function Create-ModulePsm1
            elseif ($mod["RequiredVersion"])
            {
                $importedModules += "Import-Module " + $mod["ModuleName"] + " -RequiredVersion " + $mod["RequiredVersion"] + "`r`n"
-           }
```
Comment (markcowl) [inline tools/UpdateModules.ps1]: please revert any changes here

File: tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs [hunk 5]
```diff
@@ -67,7 +67,7 @@ private static bool IsAssemblyFile(string path)
             return result;
         }
         /// <summary>
-        /// Given a set of directory paths containing PowerShell module folders, analyze the help
```
Comment (markcowl) [inline tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs]: please revert any changes here

File: tools/StaticAnalysis/BreakingChangeAnalyzer/BreakingChangeAnalyzer.cs [hunk 6]
```diff
@@ -41,7 +41,7 @@ public BreakingChangeAnalyzer()
         }
 
         /// <summary>
-        /// Given a set of directory paths containing PowerShell module folders,
```
Comment (markcowl) [inline tools/StaticAnalysis/BreakingChangeAnalyzer/BreakingChangeAnalyzer.cs]: remove any changes to this file

File: src/ResourceManager/Profile/Commands.Profile/Models/SimpleAccessToken.cs [hunk 7]
```diff
@@ -50,7 +50,7 @@ public SimpleAccessToken(IAzureAccount account, string tenantId, string tokenTyp
             }
             this.UserId = account.Id;
```
Comment (markcowl) [inline src/ResourceManager/Profile/Commands.Profile/Models/SimpleAccessToken.cs]: revert any changes here

File: src/ResourceManager/Network/Commands.Network/LoadBalancer/NewAzureLoadBalancerCommand.cs [hunk 8]
```diff
@@ -190,4 +190,4 @@ private PSLoadBalancer CreateLoadBalancer()
             return getLoadBalancer;
         }
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/LoadBalancer/NewAzureLoadBalancerCommand.cs]: revert changes here

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj [hunk 9]
```diff
@@ -554,6 +554,10 @@
       <Project>{5ee72c53-1720-4309-b54b-5fb79703195f}</Project>
       <Name>Commands.Common</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\Common\Commands.Common.Strategies\Commands.Common.Strategies.csproj">
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: revert changes here

File: src/ResourceManager/Compute/Commands.Compute/packages.config [hunk 10]
```diff
@@ -4,6 +4,11 @@
   <package id="Microsoft.Azure.KeyVault.Core" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Compute" version="17.0.0" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.KeyVault" version="2.3.0-preview" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.Network" version="15.1.0-preview" targetFramework="net452" />
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/packages.config]: we should not add packages here

File: src/ResourceManager/Common/Commands.Common.Strategies/Commands.Common.Strategies.csproj (hunk 3 above)
Comment (sergey-shandar) [reply src/ResourceManager/Common/Commands.Common.Strategies/Commands.Common.Strategies.csproj]: Should it be enough for signing?

```xml
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <OutputPath>bin\Release\</OutputPath>
    <DefineConstants>TRACE;SIGN</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <SignAssembly>true</SignAssembly>
    <DelaySign>true</DelaySign>
    <AssemblyOriginatorKeyFile>MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
  </PropertyGroup>
```

Is there anything else I should add?

File: src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs (hunk 2 above)
Comment (sergey-shandar) [reply src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs]: I've converted the project to XUnit.

File: src/ResourceManager/Compute/Commands.Compute/packages.config (hunk 10 above)
Comment (sergey-shandar) [reply src/ResourceManager/Compute/Commands.Compute/packages.config]: I removed the references.

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj (hunk 9 above)
Comment (sergey-shandar) [reply src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: done

File: src/ResourceManager/Network/Commands.Network/LoadBalancer/NewAzureLoadBalancerCommand.cs (hunk 8 above)
Comment (sergey-shandar) [reply src/ResourceManager/Network/Commands.Network/LoadBalancer/NewAzureLoadBalancerCommand.cs]: done

File: src/ResourceManager/Profile/Commands.Profile/Models/SimpleAccessToken.cs (hunk 7 above)
Comment (sergey-shandar) [reply src/ResourceManager/Profile/Commands.Profile/Models/SimpleAccessToken.cs]: done

File: tools/StaticAnalysis/BreakingChangeAnalyzer/BreakingChangeAnalyzer.cs (hunk 6 above)
Comment (sergey-shandar) [reply tools/StaticAnalysis/BreakingChangeAnalyzer/BreakingChangeAnalyzer.cs]: done

File: tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs (hunk 5 above)
Comment (sergey-shandar) [reply tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs]: done

File: tools/UpdateModules.ps1 (hunk 4 above)
Comment (sergey-shandar) [reply tools/UpdateModules.ps1]: done

File: src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs [hunk 11]
```diff
@@ -98,5 +98,7 @@ public string ResourceGroupName
 
         // Gets or sets the virtual machine zones.
         public IList<string> Zones { get; set; }
+
+        public string Fqdn { get; set; }
```
Comment (markcowl) [inline src/ResourceManager/Compute/Commands.Compute/Models/PSVirtualMachine.cs]: I tyhink this should be 'DomainName',

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/246/
signing run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-sign/34/

File: src/Common/Commands.Common/Extensions/CmdletExtensions.cs [hunk 12]
```diff
@@ -178,6 +178,30 @@ public static void SafeCopyValue<T>(this FieldInfo field, T source, T target)
...
+                if (!string.IsNullOrWhiteSpace(source.ParameterSetName))
+                {
+                    try
+                    {
+                        target.SetParameterSet(source.ParameterSetName);
+                    }
+                    catch
+                    {
+
+                    }
```
Comment (hovsepm) [inline src/Common/Commands.Common/Extensions/CmdletExtensions.cs]: would be good to at least put some info somewhere (in the debug stream?) that there were an exception.

Comment (sergey-shandar) [reply src/Common/Commands.Common/Extensions/CmdletExtensions.cs]: See this PR https://github.com/Azure/azure-powershell/pull/5086 @markcowl

File: src/ResourceManager/Common/Commands.Common.Strategies/IResourceStrategy.cs [hunk 13]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Microsoft.Azure.Commands.Common.Strategies
+{
```
Comment (hovsepm) [inline src/ResourceManager/Common/Commands.Common.Strategies/IResourceStrategy.cs]: license header missing  ;)

File: src/ResourceManager/Common/Commands.Common.Strategies/LocationExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            public DependencyLocation Visit<TModel>(ResourceConfig<TModel> config, IState state)
+                where TModel : class
+                => state.GetDependencyLocation(config);
+        }
+
+        sealed class DependencyLocation
+        {
+            public string Location { get; }
+
+            public bool IsCompulsory { get; }
```
Comment (hovsepm) [inline src/ResourceManager/Common/Commands.Common.Strategies/LocationExtensions.cs]: >IsCompulsory [](start = 24, length = 12)

does this mean IsMandatory? Maybe change the name to be consistent with parameterset attribute "Mandatory = true"

File: src/ResourceManager/Common/Commands.Common.Strategies/ResourceStrategy.cs [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            Func<TOperation, GetAsyncParams, Task<TModel>> getAsync,
+            Func<TOperation, CreateOrUpdateAsyncParams<TModel>, Task<TModel>> createOrUpdateAsync,
+            Func<TModel, string> getLocation,
+            Action<TModel, string> setLocation,
+            Func<TModel, int> createTime,
+            bool compulsoryLocation)
+            where TClient : ServiceClient<TClient>
+            => Create(
+                type,
+                name => new[] { "providers" }.Concat(headers).Concat(new[] { name }),
```
Comment (hovsepm) [inline src/ResourceManager/Common/Commands.Common.Strategies/ResourceStrategy.cs]: >headers [](start = 53, length = 7)

this should be changed to "providers". I did revert my changes due to conflicts :)

File: src/ResourceManager/Common/Commands.Common.Strategies/LocationExtensions.cs (hunk 14 above)
Comment (sergey-shandar) [reply src/ResourceManager/Common/Commands.Common.Strategies/LocationExtensions.cs]: It has different meaning. It means the resource requires all dependent resources has to have the same location. I like `IsIntrusive`. @markcowl what do you think?

**Comment by hovsepm** [comment]
> using Xunit;

license header.... yes yes... even in tests :) 

---
Refers to: src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs:1 in bfd3311. [](commit_id = bfd33116efc3f8d9539d8669946214cb2ca36241, deletion_comment = False)

File: src/ResourceManager/Compute/AzureRM.Compute.Netcore.psd1 [hunk 16]
```diff
@@ -61,7 +61,8 @@ RequiredAssemblies = 'netcoreapp2.0\AutoMapper.dll',
     'netcoreapp2.0\Microsoft.Data.Edm.dll',
     'netcoreapp2.0\Microsoft.Data.OData.dll',
     'netcoreapp2.0\Microsoft.WindowsAzure.Storage.dll',
-    'netcoreapp2.0\System.Spatial.dll'
+    'netcoreapp2.0\System.Spatial.dll', 
+    '.\Microsoft.Azure.Commands.Common.Strategies.dll'
```
Comment (hovsepm) [inline src/ResourceManager/Compute/AzureRM.Compute.Netcore.psd1]: >.\ [](start = 5, length = 2)

shouldn't this as well reference strategies.dll in the netcoreapp2.0 folder? btw there were some issues with .\ path on linux on one of the versions of .Net Core PS.

Comment (sergey-shandar) [reply src/ResourceManager/Compute/AzureRM.Compute.Netcore.psd1]: @hovsepm I think, you are right. @twitchax please, confirm.

File: src/ResourceManager/Common/Commands.Common.Strategies/LocationExtensions.cs (hunk 14 above)
Comment (sergey-shandar) [reply src/ResourceManager/Common/Commands.Common.Strategies/LocationExtensions.cs]: @hovsepm Ok. we don't have a better name for now. I commented this property. The property is internal so we will have a chance to change it if we have a better name.

File: src/Common/Commands.Common/Extensions/CmdletExtensions.cs (hunk 12 above)
Comment (markcowl) [reply src/Common/Commands.Common/Extensions/CmdletExtensions.cs]: This is bootstrapping code that cannot fail, so we should not use the PowerShell ICommandRuntime

File: src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+// ----------------------------------------------------------------------------------
+
+using Xunit;
+
+namespace Microsoft.Azure.Commands.Common.Strategies.UnitTest
+{
+    public class TimeSlotTest
+    {
+        [Fact]
+        public void AddTest()
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs]: This should be a check-in tess, we use traits to select tests used for check-in, see any of the test projects 
```[Trait(Category.AcceptanceType, Category.CheckIn)]```

Comment (sergey-shandar) [reply src/ResourceManager/Common/Commands.Common.Strategies.UnitTest/TimeSlotTest.cs]: fixed.

**Comment by markcowl** [comment]
on demand: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/255/

**Pull request merged**